rust/src/websockets/websockets.rs

```diff
@@ -0,0 +1,268 @@
+        Self {
+            tx_data: AppLayerTxData::for_direction(direction),
+            ..Default::default()
+        }
+    }
+}
+
+impl Transaction for WebSocketsTransaction {
+    fn id(&self) -> u64 {
+        self.tx_id
+    }
+}
+
+#[derive(Default)]
+pub struct WebSocketsState {
+    state_data: AppLayerStateData,
+    tx_id: u64,
+    transactions: VecDeque<WebSocketsTransaction>,
+}
+
+impl State<WebSocketsTransaction> for WebSocketsState {
+    fn get_transaction_count(&self) -> usize {
+        self.transactions.len()
+    }
+
+    fn get_transaction_by_index(&self, index: usize) -> Option<&WebSocketsTransaction> {
+        self.transactions.get(index)
+    }
+}
+
+impl WebSocketsState {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    // Free a transaction by ID.
+    fn free_tx(&mut self, tx_id: u64) {
+        let len = self.transactions.len();
+        let mut found = false;
+        let mut index = 0;
+        for i in 0..len {
+            let tx = &self.transactions[i];
+            if tx.tx_id == tx_id + 1 {
+                found = true;
+                index = i;
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&WebSocketsTransaction> {
+        self.transactions.iter().find(|tx| tx.tx_id == tx_id + 1)
+    }
+
+    fn new_tx(&mut self, direction: Direction) -> WebSocketsTransaction {
+        let mut tx = WebSocketsTransaction::new(direction);
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        return tx;
+    }
+
+    fn parse(
+        &mut self, stream_slice: StreamSlice, direction: Direction, flow: *const Flow,
+    ) -> AppLayerResult {
+        let input = stream_slice.as_slice();
+        let mut start = input;
+        while !start.is_empty() {
+            match parser::parse_message(start) {
+                Ok((rem, pdu)) => {
+                    let _pdu = Frame::new(
+                        flow,
+                        &stream_slice,
+                        start,
+                        (start.len() - rem.len() - pdu.payload.len()) as i64,
+                        WebSocketsFrameType::Header as u8,
+                    );
+                    let _pdu = Frame::new(
+                        flow,
+                        &stream_slice,
+                        start,
+                        (start.len() - rem.len()) as i64,
+                        WebSocketsFrameType::Pdu as u8,
+                    );
+                    start = rem;
+                    let mut tx = self.new_tx(direction);
+                    tx.pdu = pdu;
+                    //TODOws should we reassemble/stream payload data ?
+                    self.transactions.push_back(tx);
+                }
+                Err(nom::Err::Incomplete(_)) => {
+                    // Not enough data. just ask for one more byte.
```

Comment:
as the header seems to have a payload_len, could we use that here?

Also what is the max payload size?

Reply:
A frame, per spec can be up to 9,223,372,036,854,775,807 bytes I believe.

Reply:
but I guess it means we need to consume all payload data directly, even if incomplete. Process them the same as other streaming buffers like http body and file.data

Reply:
> frames the size of the universe, of course

In practice, most implementations don't allow extremely large payloads, I think 16kb even being on the extreme..  But we of course still need to deal with universe sized frames somehow.